Create a custom navigator to handle top level fragments navigation

hichamboushaba opened the pull request:
First part of #3411, the changes included are:

1. Create a custom navigator that handles navigation of toplevel fragments, while keeping their state.
2. Add the top level fragments to the navgraph.
3. Update MainActivity and MainBottomNavigationView to use the new navigator.
4. Update logic of different top level fragments to avoid accessing the views in `onSaveInstanceState`, the views are available currently only between `onCreateView` and `onDestroyView`, and as fragments are detached, `onSaveInstanceState` may be called before `onDestroyView`.
5. Update handling of toolbar expansion after navigation events: 439b68c46029c79495f095d07145ae42ddbc8849

A known issue: the bottom navigation bar may appear in some child fragments, this will be handled in the next PR.

Testing:
Test app's navigation, and confirm that it's working as expected.

Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

hichamboushaba committed:
Add custom navigator to handle top level navigation

hichamboushaba committed:
Add top level fragments to nav_graph

hichamboushaba committed:
Update the navigation to use the new navigator

hichamboushaba committed:
Fix issues related to screen rotation of MyStoreFragment

hichamboushaba committed:
Fix issues related to screen rotation of OrderListFragment

hichamboushaba committed:
Fix ProductListFragment

hichamboushaba committed:
Fix ReviewListFragment

hichamboushaba committed:
Remove the unused container layout

hichamboushaba committed:
Fix getTopLevelFragment impl

anitaa1990 commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/mystore/MyStoreFragment.kt:
```diff
@@ -94,6 +92,21 @@ class MyStoreFragment : TopLevelFragment(R.layout.fragment_my_store),
     private val myStoreDateBar
         get() = binding.myStoreStats.myStoreDateBar
 
+    private var isEmptyViewVisible: Boolean = false
```
np: Would it not make more sense to use `binding.emptyView.visibility == View.VISIBLE` here instead of using a `var` and defining the `isEmptyViewVisible` in a separate method?

hichamboushaba replied:
No, that's the point of the change, we can't access `binding` after the view destroying, so we need to keep this as a separate variable.
Once we convert this to MVVM we won't need it normally, as the ViewModel won't be destroyed.

anitaa1990 commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/mystore/MyStoreFragment.kt:
```diff
@@ -79,11 +79,9 @@ class MyStoreFragment : TopLevelFragment(R.layout.fragment_my_store),
-            it.setId(R.id.stats_tab_layout)
-        }
-    }
+    private var _tabLayout: TabLayout? = null
+    private val tabLayout
+        get() = _tabLayout!!
```
I noticed that the tabLayout style has been removed from here (`R.attr.scrollableTabStyle`). Is that intentional?

So it looks like this now:
<img src="https://user-images.githubusercontent.com/22608780/104413781-f93d5880-5594-11eb-9db3-7984351fcbcb.png" width="300"/>

When it should look like this?
<img src="https://user-images.githubusercontent.com/22608780/104413827-0eb28280-5595-11eb-9100-ce2175a0c9a0.png" width="300"/>

hichamboushaba replied:
Nice catch 👍

anitaa1990 reviewed (commented):
Nice work on this @hichamboushaba! I've left a few comments in the code for your review. There are also some changes that I see in the UX when compared to `develop` (For instance, the bottom nav should not be displayed in any of the child fragments, the Add product button is currently visible in all the product sub screens). But I assume those will be taken care of in another PR?

Also there is a lint error in the PR, that's causing CircleCI to fail. Not sure if any of the @woocommerce/android-developers would like to take a look at the PR as well?

hichamboushaba commented:
> Nice work on this @hichamboushaba! I've left a few comments in the code for your review. There are also some changes that I see in the UX when compared to `develop` (For instance, the bottom nav should not be displayed in any of the child fragments, the Add product button is currently visible in all the product sub screens). But I assume those will be taken care of in another PR?
> 
> Also there is a lint error in the PR, that's causing CircleCI to fail. Not sure if any of the @woocommerce/android-developers would like to take a look at the PR as well?  
> 

Thanks @anitaa1990, the PR is still a draft because I've been working on those issues 🙂, I'll mark it as ready for review after I finish them.

hichamboushaba committed:
Move the toolbar expansion logic to fragment lifecycle observer

As when navigating to a different fragment, the view will be destroyed, updating the toolbar status after view creation is what we need.
In addition to that, we can't access `isScrolledToTop` before the view is created.

hichamboushaba committed:
Update OrderListFragment logic for `isScrolledToTop`

hichamboushaba committed:
Hide products FAB when destroying the ProductListFragment view

hichamboushaba committed:
Make the tablayout scrollable in MyStoreFragment

hichamboushaba committed:
Fix ktlint issues

hichamboushaba commented:
The PR is ready now for review.
The issue of showing the bottom nav bar in some child fragments will be handled in the next PR when the TopLevelFragment logic is cleaned.

hichamboushaba committed:
Remove the unused key

hichamboushaba committed:
Add a comment explaining the need to post the call on the view handler

hichamboushaba commented:
I'm adding you as a reviewer @0nko, as Anitaa will be off the next two days.

0nko reviewed (commented):
I played around with the navigation, just randomly moved back and forth between top-level fragments, rotating the device (with `Don't keep activities` enabled) and I got this crash:
```
2021-01-14 12:12:06.451 2419-2419/com.woocommerce.android.prealpha E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.woocommerce.android.prealpha, PID: 2419
    kotlin.KotlinNullPointerException
        at com.woocommerce.android.ui.orders.list.OrderListFragment.getTabLayout(OrderListFragment.kt:114)
        at com.woocommerce.android.ui.orders.list.OrderListFragment.addTabLayoutToAppBar(OrderListFragment.kt:766)
        at com.woocommerce.android.ui.orders.list.OrderListFragment.onReturnedFromChildFragment(OrderListFragment.kt:279)
        at com.woocommerce.android.ui.main.MainActivity.onDestinationChanged(MainActivity.kt:461)
        at androidx.navigation.NavController.dispatchOnDestinationChanged(NavController.java:498)
        at androidx.navigation.NavController.navigate(NavController.java:1097)
        at androidx.navigation.NavController.navigate(NavController.java:935)
        at androidx.navigation.NavController.navigate(NavController.java:868)
        at com.woocommerce.android.ui.main.MainBottomNavigationView.onNavigationItemSelected(MainBottomNavigationView.kt:187)
        at com.google.android.material.bottomnavigation.BottomNavigationView$1.onMenuItemSelected(BottomNavigationView.java:243)
        at androidx.appcompat.view.menu.MenuBuilder.dispatchMenuItemSelected(MenuBuilder.java:834)
        at androidx.appcompat.view.menu.MenuItemImpl.invoke(MenuItemImpl.java:158)
        at androidx.appcompat.view.menu.MenuBuilder.performItemAction(MenuBuilder.java:985)
        at com.google.android.material.bottomnavigation.BottomNavigationMenuView$1.onClick(BottomNavigationMenuView.java:127)
        at android.view.View.performClick(View.java:6294)
        at android.view.View$PerformClick.run(View.java:24770)
        at android.os.Handler.handleCallback(Handler.java:790)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:164)
        at android.app.ActivityThread.main(ActivityThread.java:6494)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:438)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:807)
```

hichamboushaba commented:
@0nko this crash will be solved in the third PR when I get to cleaning up the logic of top level fragments. What happens is that we try to call `onReturnedFromChildFragment` in `onDestinationChanged` callback, but as it's called before creating the view of the fragment, it returns null.
Normally with the new navigation logic, we won't need the callback `onReturnedFromChildFragment` as it mostly delt with updating the menu and the view, because the top-level fragment was still active.

0nko reviewed (approved):
Solid work, @hichamboushaba! 👍
